repose-aggregator/components/filters/openstack-identity-v3-filter/src/main/scala/org/openrepose/filters/openstackidentityv3/OpenStackIdentityV3Handler.scala

```diff
@@ -87,6 +86,19 @@ class OpenStackIdentityV3Handler(identityConfig: OpenstackIdentityV3Config, iden
       response.addHeader(HttpHeaders.RETRY_AFTER, retry)
     }
 
+    implicit val catalogWrites = new Writes[ServiceForAuthenticationResponse] {
+      def writes(catalog: ServiceForAuthenticationResponse) = JsObject(Seq(
+        "endpoints" -> JsArray(catalog.endpoints.map(endpoint => JsObject(
+          Seq("id" -> JsString(endpoint.id)) ++
+            endpoint.name.map(name => Seq("name" -> JsString(name))).getOrElse(Seq()) ++
+            endpoint.interface.map(interface => Seq("interface" -> JsString(interface))).getOrElse(Seq()) ++
+            endpoint.region.map(region => Seq("region" -> JsString(region))).getOrElse(Seq()) ++
+            Seq("url" -> JsString(endpoint.url)) ++
+            endpoint.service_id.map(serviceId => Seq("service_id" -> JsString(serviceId))).getOrElse(Seq()))))) ++
+        catalog.id.map(id => Seq("id" -> JsString(id))).getOrElse(Seq()) ++
+        catalog.name.map(name => Seq("name" -> JsString(name))).getOrElse(Seq()))
+    }
```

Comment:
I should just wait for you to finish with your changes, but...

1. Sounds legit.

2. Right, right, so while it's in the same section in the docs, Play JSON supports class construction and function construction. So rather than writing `JsObject(stuff)` you could write `Json.obj(stuff)`. Mostly, this should allow you to replace the `Seq`s with calls to functions which take varargs.

3. Well, technically it would be `Seq[(String, JsValue)]` vs `Seq.empty[(String, JsValue)]`. In both cases, though, Scala's type system should be able to figure out what you want without being explicit about the types. So really, it would be `Seq()` vs `Seq.empty`.

4. So there is absolutely a conversion from `String` to `JsString`, but the type system has to know to do that conversion. In your case, you can coerce the type system by writing something like `Seq[(String, JsString)]("foo" -> "bar")`. That should invoke the implicit conversion. Of course, if you collapse some of the other stuff here, you shouldn't even need to be explicit about the types. The type system will handle a lot of it for you.

4a. Ah. Totally didn't realize the old behavior left stuff out. Man, that's kind of gross.

5. Cool, cool.